GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -559,39 +631,49 @@ private RevisionGraphRevision[] BuildOrderedNodesCache(int currentRowIndex)
                 {
-                    // Reset the reorder flag and the orderedUntilScore. This makes sure it isn't marked dirty before we even got to
-                    // rebuilding it.
+                    // Reset flags to sure it isn't marked dirty before we even got to rebuilding it.
                     _orderedUntilScore = int.MinValue;
-                    _reorder = false;
+                    _reorderNodeCache = false;
 
-                    // Use a local variable, because the cached list can be reset.
-                    RevisionGraphRevision[] localOrderedNodesCache = _nodes.ToArray();
-                    Array.Sort(localOrderedNodesCache, (x, y) => x.Score.CompareTo(y.Score));
-                    _orderedNodesCache = localOrderedNodesCache;
-                    if (localOrderedNodesCache.Length > 0)
+                    _orderedNodesCache = _revisionByObjectId.Values.OrderBy(n => n.Score).ToImmutableArray();
+                    if (_orderedNodesCache.Length > 0)
                     {
-                        _orderedUntilScore = localOrderedNodesCache[^1].Score;
+                        _orderedUntilScore = _orderedNodesCache[^1].Score;
                     }
 
-                    return localOrderedNodesCache;
+                    return _orderedNodesCache;
                 }
-                catch (ArgumentException ex) when (_reorder && ex.Message.Contains("IComparer.Compare()"))
+                catch (ArgumentException ex) when (_reorderNodeCache && ex.Message.Contains("IComparer.Compare()"))
                 {
                     // ignore and try again
                 }
             }
         }
 
         /// <summary>
-        /// Mark the cache as invalid if the node scores have been changed.
+        /// Mark the node and row cache as invalid if the node scores in the ordered node cache have been changed.
         /// </summary>
         /// <param name="minScore">The minimum score of the modified nodes.</param>
-        private void MarkCacheAsInvalidIfNeeded(int minScore)
+        /// <returns><see langword="true"/> if the row cache is invalidated; otherwise <see langword="false"/>.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool MarkCacheAsInvalidIfNeeded(int minScore)
         {
             if (minScore <= _orderedUntilScore)
             {
-                _reorder = true;
+                _reorderNodeCache = true;
             }
+
+            // The row cache is expensive to build, find if the built index is affected.
+            // This is not expected to be a common issue (primarily inserting artificial,
+            // topo order and tests).
+            // Not fully optimized, invalidate the complete cache instead of only the rows after the index.
+            ImmutableArray<RevisionGraphRow> localOrderedRowCache = _orderedRowCache;
+            if (localOrderedRowCache.Length > 0 && minScore <= localOrderedRowCache[^1].Revision.Score)
```

Comment:
Use `_orderedUntil`?

Reply:
RowCache would have a separate variable, but the node cache variable was removed. No change

Reply:
row vs. node cache (it was too late in the evening)